File: nnerik/day03.py
import re

# Read input
input = open(0).read()


# Part 1
instructions = re.finditer(r"mul\((\d{1,3}),(\d{1,3})\)", input)
total = 0
for pair in instructions:
    a, b = pair.groups()
    total += int(a) * int(b)

print("Part 1:", total)


# Part 2
total = 0
instructions = re.finditer(r"(do)\(\)|(don't)\(\)|mul\((\d{1,3}),(\d{1,3})\)", input)

do = True
for instruction in instructions:
    if instruction.group(1) == "do":
        do = True
        continue
    elif instruction.group(2) == "don't":
        do = False
        continue
    elif do:
        a, b = instruction.group(3, 4)
        total += int(a) * int(b)

print("Part 2:", total)

File: fredriv/day03/day03.scala
import scala.io.Source

@main def day03 =
    val input = Source.fromFile("input.txt").getLines.mkString("\n")

    val mulRegex = """mul\((\d+),(\d+)\)""".r
    val products = for m <- mulRegex.findAllMatchIn(input) yield
        m.group(1).toInt * m.group(2).toInt
    println("Part 1: " + products.sum)

    val doDontMulRegex = """mul\((\d+),(\d+)\)|do\(\)|don't\(\)""".r
    var enabled = true
    var total = 0
    for m <- doDontMulRegex.findAllMatchIn(input) do
        if m.matched == "do()" then enabled = true
        else if m.matched == "don't()" then enabled = false
        else if enabled then total += m.group(1).toInt * m.group(2).toInt
    println("Part 2: " + total)
